### PR TITLE
Overlay - Moving Around & Clipping 

### PR DESCRIPTION
**Pull request by alexstyl**
This PR allows the Overlay to be dragged around. The Overlay can be moved only vertically and it is clipped on the edges of the screen.

### What's done so far
- Draw Overlay from a Service
- Overlay Screen edge Clipping

### Coming Up Next (In different PRs)

- The proper layout of the overlay
- Overlay Interactions

overlay_clip

This PR continues the work of: Overlay - Foreground Service #2

### CODE REVIEW

File: app/src/main/java/com/alexstyl/gif/util/Utils.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.alexstyl.gif.util;
```
Comment (alexstyl) [inline app/src/main/java/com/alexstyl/gif/util/Utils.java]: This needs to be moved to `com.alexstyl.gif`. No need to have a package of utils

Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/util/Utils.java]: indeed

File: app/src/main/java/com/alexstyl/gif/util/Utils.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import android.content.res.Resources;
+
+public class Utils {
```
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/util/Utils.java]: remember this https://github.com/alexstyl/project-gif/pull/2#discussion_r49060681 :wink:

File: app/src/main/java/com/alexstyl/gif/util/CombinationTouchListener.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.alexstyl.gif.util;
+
+import android.view.MotionEvent;
+import android.view.View;
+
+import java.util.HashSet;
+
+final public class CombinationTouchListener implements View.OnTouchListener {
```
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/util/CombinationTouchListener.java]: `TouchListnenerCollection`?

File: app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+import android.content.Context;
+import android.view.MotionEvent;
+import android.view.View;
+import android.view.WindowManager;
+
+public class DragVerticallyTouchListener implements View.OnTouchListener {
+
+    private int yDelta;
+    private final WindowManager windowManager;
```
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java]: final fields above variables

File: app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public static DragVerticallyTouchListener newInstance(Context context) {
+        WindowManager manager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
+        return new DragVerticallyTouchListener(manager);
+    }
+
+    @Override
+    final public boolean onTouch(View view, MotionEvent event) {
+        final int Y = (int) event.getRawY();
```
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java]: don't think capital letters are a convention in java :tongue:

Comment (alexstyl) [inline app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java]: that is true. I have no idea what happened there.

File: app/src/main/java/com/alexstyl/gif/util/Utils.java (hunk 1 above)
Comment (alexstyl) [inline app/src/main/java/com/alexstyl/gif/util/Utils.java]: I deleted the class completely and moved the methods inside the one and only class that were using them

**Comment by alexstyl** [comment]
All comments addressed

**Comment by xrigau** [comment]
LGTM :+1:

File: app/src/main/java/com/alexstyl/gif/overlay/ClipToVerticalEdgesTouchListener.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import android.view.WindowManager;
+
+class ClipToVerticalEdgesTouchListener implements View.OnTouchListener {
+
+    private final OrientationChangedChecker orientationChangedChecker;
+    private final WindowManager windowManager;
+    private final Resources resources;
+    private final Point screenBoundaries = new Point();
+    private int topBoundary;
+    private int bottomBoundary;
```
Comment (blundell) [inline app/src/main/java/com/alexstyl/gif/overlay/ClipToVerticalEdgesTouchListener.java]: personally I like to separate all 3 of these as they are all fields for different reasons
```java
     private final Point screenBoundaries = new Point();

     private final OrientationChangedChecker orientationChangedChecker;
     private final WindowManager windowManager;
     private final Resources resources;

     private int topBoundary;
     private int bottomBoundary;
```

File: app/src/main/java/com/alexstyl/gif/overlay/ClipToVerticalEdgesTouchListener.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+            if (isClippingTop(layoutParams, viewHeight)) {
+                layoutParams.y = 0;
+            } else if (isClippingBottom(layoutParams, viewHeight)) {
+                layoutParams.y = bottomBoundary - viewHeight;
+            }
+
+            view.setLayoutParams(layoutParams);
+
+            windowManager.updateViewLayout(view, layoutParams);
```
Comment (blundell) [inline app/src/main/java/com/alexstyl/gif/overlay/ClipToVerticalEdgesTouchListener.java]: is this class doing two things? 1, calculating where the touch point is for the clipping to vertical edges and 2) updating the views position.   Perhaps this class should expose an interface that gives back the new y position - and another class should be in charge of updating the view?

File: app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                yDelta = y - lParams.y;
+                break;
+            case MotionEvent.ACTION_POINTER_DOWN:
+                break;
+            case MotionEvent.ACTION_POINTER_UP:
+                break;
+            case MotionEvent.ACTION_MOVE:
+                lParams.y = y - yDelta;
+                view.setLayoutParams(lParams);
+                windowManager.updateViewLayout(view, lParams);
```
Comment (blundell) [inline app/src/main/java/com/alexstyl/gif/overlay/DragVerticallyTouchListener.java]: same here

File: app/src/main/java/com/alexstyl/gif/overlay/OrientationChangedChecker.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package com.alexstyl.gif.overlay;
+
+import android.content.Context;
+import android.content.res.Configuration;
+
+public class OrientationChangedChecker {
```
Comment (blundell) [inline app/src/main/java/com/alexstyl/gif/overlay/OrientationChangedChecker.java]: :+1:

**Comment by alexstyl** [comment]
@blundell @xrigau All comments addressed. Changed the implementation a bit.

File: app/src/main/java/com/alexstyl/gif/Notifier.java [hunk 10]
```diff
@@ -16,6 +17,9 @@ public static Notifier newInstance(Context context) {
         this.context = context;
     }
 
+    @SuppressLint("StringFormatInvalid")
+    // the contentTitle is linking to a string with no arguments and Lint is complaining.
+    // In fact, that string links to a strings that actually links to a string with arguments
```
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/Notifier.java]: is there an issue in android's issue tracker already? if there is, it'd be nice to add the link here, if there isn't you could create one :1234:

**Comment by alexstyl** [comment]
@blundell @xrigau It's been a while... Is there any more issues with this PR that needs handling?

**Comment by blundell** [comment]
screen shot 2016-09-05 at 08 17 44

not from me, but you've kind of missed an obvious question :-)

**Comment by alexstyl** [comment]
I'll resolve it outside of the PR. No matter the outcome it can be done without affecting the merging of this PR, right? :)

**Comment by blundell** [comment]
yeah of course, it's just I don't merge things with questions that aren't answered :-)

File: app/src/main/java/com/alexstyl/gif/Notifier.java (hunk 10 above)
Comment (alexstyl) [inline app/src/main/java/com/alexstyl/gif/Notifier.java]: @xrigau I'll have a look and post the result here. Feel free to merge if this is the only issue 🙆

**Comment by blundell** [comment]
LGTM @xrigau can converse with you when he's back from hols 👍

File: app/src/main/java/com/alexstyl/gif/Notifier.java (hunk 10 above)
Comment (xrigau) [inline app/src/main/java/com/alexstyl/gif/Notifier.java]: yeah known issue https://code.google.com/p/android/issues/detail?id=201812

**Comment by xrigau** [comment]
👍  good from me

**Pull request merged**